app/core/Ledger/Ledger.ts

```
@@ -96,31 +78,18 @@ export const unlockLedgerDefaultAccount = async (): Promise<{
   balance: string;
 }> => {
   const keyringController = Engine.context.KeyringController;
-  const preferencesController = Engine.context.PreferencesController;
   const keyring = await getLedgerKeyring();
-  const oldAccounts = await keyringController.getAccounts();
-  await keyringController.addNewAccountForKeyring(keyring);
-  const newAccounts = await keyringController.getAccounts();
-
-  keyringController.updateIdentities(newAccounts);
-  newAccounts.forEach((address: string) => {
-    if (!oldAccounts.includes(address)) {
-      if (keyringController.setAccountLabel) {
-        // The first ledger account is always returned.
-        keyringController.setAccountLabel(address, `${keyring.getName()} 1`);
-      }
-      preferencesController.setSelectedAddress(address);
-    }
-  });
-  await keyringController.persistAllKeyrings();
-
+  try {
+    await keyringController.addNewAccountForKeyring(keyring);
+  } catch (e) {
+    console.error(e);
+  }
   const address = await keyring.getDefaultAccount();
   return {
     address,
     balance: `0x0`,
   };
```

What happens if the account fails to be added as we ignore the error?

Moreover, `keyringController.addNewAccountForKeyring` returns the added account: could we use that instead of calling `getDefaultAccount`?

---

First of all, `unlockLedgerDefaultAccount` method has been used in `LedgerConnect` (during bluetooth paired with ledger and then import ledger account and message signing flow), and then `LedgerTransactionConfirmationModal' for all ledger transaction interaction.

Previously, `keyringController.addNewAccountForKeyring()` doesn't throw any exception if keyring already exists in keyringController, but for somehow, after `KeyringController v8.1 upgrade`, `addNewAccountForKeyring()` start to throw exception during transaction flow and signing flow and transaction flow.  the above code has been introduce to quickly fix the issue during `signing` and `transactioni` flow.

However, i agreed that above try catch block is not robust, currently i try to find a way to avoid running `addNewAccountForKeyring()`, i will check whether selectedAccount in metamask is  ledger defaultAccount, if yes, which mean the operation flow is `ledger message signing` or `ledger transaction flow`.  if selectedAccount is not ledger defaultAccount, then, `unlockLedgerDefaultAccount` will be called by pairing ledger and import ledger account. then `addNewAccountForKeyring()` should be run.

@mikesposito To answer your question, i will remove the `try catch` block above, so if during pairing ledger and import ledger account flow failed to add account. then ` await keyringController.addNewAccountForKeyring(keyring);` will directly throw exception for system to catch.

---

have modified the `unlockLedgerDefaultAccount` method to include extra flag as argument to check whether the call to this method require to import a new account or not.  i have removed the `try catch` block in that method as well so that `addNewAccountForKeyring` can throw exception if account fails to be added.

---

hi, @mikesposito yea, the exception will bubble up and catch in the ledger connect screen when adding account,  that is previous code did.